### PR TITLE
FT connector: add whitelisting feature

### PR DESCRIPTION
**Pull request by karim-en**
Add whitelist support to NEAR -> ETH connector

### Near side:
The whitelist mode is enabled by default on init (`is_whitelist_mode_enabled = ture`)
Call `token-locker::set_token_whitelist_mode(mode)` to add a token to the whitelist and provide one of the modes:
-  Blocked: the deposit is not allowed for this token. (this is the default behavior)
-  CheckToken: the deposit is allowed for this token by any account.
-  CheckAccountAndToken: the deposit is allowed for this token by accounts from the whitelist.

Call `token-locker::addAccountToWhitelist(token, account)` to add an account to the white list.

### Eth side:
The whitelist mode is enabled by default on deploy.

Call `BridgeTokenFactory::enable_whitelist_mode()` or `BridgeTokenFactory::disable_whitelist_mode()` to enable or disable the whitelist mode. (Only allowed by `DEFAULT_ADMIN_ROLE`)

To add a token to the whitelist we need to call `BridgeTokenFactory::setTokenWhitelistMode(mode)` and provide one of the modes: (Only allowed by `WHITELIST_ADMIN_ROLE`)
-  Blocked: the withdrawal is not allowed for this token. (this is the default behavior)
-  CheckToken: the withdrawal is allowed for this token by any account.
-  CheckAccountAndToken: the withdrawal is allowed for this token only by accounts from the whitelist.

Call `BridgeTokenFactory::addAccountToWhitelist(token, account)` to add an account to the whitelist. (Only allowed by `WHITELIST_ADMIN_ROLE`)

### CODE REVIEW

File: erc20-bridge-token/contracts/BridgeTokenFactory.sol [hunk 1]
```diff
@@ -10,13 +11,12 @@ import "rainbow-bridge-sol/nearprover/contracts/INearProver.sol";
...
 import "rainbow-bridge-sol/nearbridge/contracts/Borsh.sol";
 
-
 import "./IProofConsumer.sol";
 import "./BridgeToken.sol";
 import "./BridgeTokenProxy.sol";
 
 contract BridgeTokenFactory is  AccessControlUpgradeable, PausableUpgradeable{
-
+    enum WhitelistMode { NotInited, Blocked, CheckToken, CheckAccountAndToken }
```
Comment (sept-en) [inline erc20-bridge-token/contracts/BridgeTokenFactory.sol:19]:
```suggestion
    enum WhitelistMode {
        NotInited,
        Blocked,
        CheckToken,
        CheckAccountAndToken
    }
    
```

File: erc20-bridge-token/contracts/BridgeTokenFactory.sol [hunk 2]
```diff
@@ -154,4 +161,35 @@ contract BridgeTokenFactory is  AccessControlUpgradeable, PausableUpgradeable{
...
+    function _check_whitelist_token(string memory token, address account) internal view {
+        if (!_is_whitelist_mode_enabled) {
+            return;
+        }
+        
+        WhitelistMode token_mode = _whitelist_tokens[token];
+        require(token_mode != WhitelistMode.NotInited, "ERR_NOT_INITED_WHITELIST_TOKEN");
+
+        if (token_mode == WhitelistMode.CheckAccountAndToken) {
+            require(_whitelist_accounts[abi.encodePacked(token, account)], "ERR_ACCOUNT_NOT_IN_WHAITELIST");
```
Comment (sept-en) [inline erc20-bridge-token/contracts/BridgeTokenFactory.sol:190]:
```suggestion
            require(_whitelist_accounts[abi.encodePacked(token, account)], "ERR_ACCOUNT_NOT_IN_WHITELIST");
```

File: erc20-bridge-token/contracts/BridgeTokenFactory.sol [hunk 3]
```diff
@@ -154,4 +161,35 @@ contract BridgeTokenFactory is  AccessControlUpgradeable, PausableUpgradeable{
...
+        } else if (token_mode == WhitelistMode.Blocked) {
+            revert("ERR_WHAITELIST_TOKEN_BLOCKED");
```
Comment (sept-en) [inline erc20-bridge-token/contracts/BridgeTokenFactory.sol:192]:
```suggestion
            revert("ERR_WHITELIST_TOKEN_BLOCKED");
```

File: token-locker/src/whitelist.rs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    pub fn check_whitelist_token(&mut self, token: AccountId, account: AccountId) {
+        if !self.is_whitelist_mode_enabled {
+            return;
+        }
+
+        let token_whitelist_mode = self
+            .whitelist_tokens
+            .get(&token)
+            .unwrap_or_else(|| env::panic_str("The token is not exist in the whitelist"));
```
Comment (sept-en) [inline token-locker/src/whitelist.rs:36]:
```suggestion
            .unwrap_or_else(|| env::panic_str("The token is not whitelisted"));
```

File: token-locker/src/whitelist.rs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+        match token_whitelist_mode {
+            WhitelistMode::CheckAccountAndToken => {
+                let token_account_key = format!("{}:{}", token, account);
+                assert!(
+                    self.whitelist_accounts.contains(&token_account_key),
+                    "The token:account key is not exist in the whitelist"
+                );
```
Comment (sept-en) [inline token-locker/src/whitelist.rs:44]:
```suggestion
                assert!(
                    self.whitelist_accounts.contains(&token_account_key),
                    "{}",
                    format!(
                        "The {} key does not exist in the whitelist",
                        token_account_key
                    ),
                );
```

File: token-locker/src/whitelist.rs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    pub fn set_token_whitelist_mode(&mut self, token: AccountId, mode: WhitelistMode) {
+        assert_self();
+        self.whitelist_tokens.insert(&token, &mode);
+    }
+
+    pub fn add_account_to_whitelist(&mut self, token: AccountId, account: AccountId) {
+        assert_self();
+        assert!(
+            self.whitelist_tokens.get(&token).is_some(),
+            "The token mode should be set first"
```
Comment (sept-en) [inline token-locker/src/whitelist.rs:22]:
```suggestion
            "The whitelisted token mode is not set",
```

File: token-locker/src/whitelist.rs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        );
+        let token_account_key = format!("{}:{}", token, account);
```
Comment (sept-en) [inline token-locker/src/whitelist.rs:24]: Please move `token_account_key` calculation to a separate function as it's used twice

**Review by sept-en — CHANGES_REQUESTED** [review]

File: token-locker/src/whitelist.rs (hunk 7 above)
Comment (karim-en) [reply token-locker/src/whitelist.rs]: Fixed 3c8415c85f4f85a003ca9ee1ff925f2611807a09

**Review by sept-en — APPROVED** [review]

**Pull request merged**